workers/unity/Packages/io.improbable.gdk.tools/CodeGenDocs/2-write-a-generator.md

```diff
@@ -0,0 +1,270 @@
+{
+    return CodeWriter.Populate(cw =>
+    {
+        cw.UsingDirectives(
+            "System",
+            "System.Collections.Generic"
+        );
+
+        cw.Namespace("Your.Own.Namespace", ns =>
+        {
+            //define namespace contents
+        });
+    });
+}
+```
+
+## CodeWriter: Namespaces
+
+You can use the following methods when constructing a namespace:
+
+- `Annotate`
+- `Text`
+- `TextList`
+- `Line`
+- `CustomScope`
+- `Enum`
+- `Type`
+
+## CodeWriter: Structs and classes
+
+Structs and classes can be defined using the `Type` method where available.
+
+You can use the following familiar methods when constructing structs and classes:
+
+- `Annotate`
+- `Text`
+- `TextList`
+- `Line`
+- `CustomScope`
+- `Enum`
+- `Type`
+
+In addition, the following two methods are exposed:
+
+- `Initializer`
+- `Method`
+
+## Enums
+
+You can use the following methods when constructing enums:
+
+- `Member`
+- `Members`
+
+## CodeWriter: Other scoped bodies
+
+A scoped body is a generic body of code that is either a method or can lie within a method scope. Constructs of this kind are:
+
+- `CustomScope`
+- `Method`
+- `Loop`
+
+You can use the following methods when defining a scoped body:
+
+- `Text`
+- `TextList`
+- `Line`
+- `CustomScope`
+- `Initializer`
+
+The following methods are also exposed:
+
+- `Loop`
+- `If/ElseIf/Else`
+- `Try/Catch/Finally`
+- `Return/Return(x)`
+- `ProfilerStart(x)`
+- `ProfilerEnd()`
+
+> Note: the above methods are not exposed at top level types/namespaces/enums as they can only be used within method bodies.
+
+### Custom scope
+
+A custom scope is a a scope that can be defined with or without a declaration. For example:
+
+```csharp
+using (var thing = new SomethingDisposable())
+{
+    //scope with a declaration
+}
+
+{
+    //scope without declaration
+}
+```
+
+You can use this to define any un-annotated code construct, but it is recommended for constructs that don't already have a helper API.
+
+### Methods
+
+Methods are a special type of custom scope block that can also be annotated.
+
+### Loops
+
+Loops are a special type of custom scope block that also expose `Continue` and `Break` methods.
+
+The following examples would all return identical generated code:
+
+```csharp
+m.Loop("foreach (var x in xs)", e =>
+{
+    e.If("x == 0", then =>
+    {
+        then.Continue();
+    })
+    .ElseIf("x > 5", then =>
+    {
+        then.Break();
+    });
+
+    //logic
+});
+```
+
+```csharp
+m.Loop("foreach (var x in xs)", e =>
+{
+    e.If("x == 0", then =>
+    {
+        then.Line("continue;");
+    })
+    .ElseIf("x > 5", then =>
+    {
+        then.Line("break;");
+    });
+
+    //logic
+});
+```
+
+```csharp
+m.Loop("foreach (var x in xs)", e =>
+{
+    e.If("x == 0", () => new[]
+    {
+        "continue;"
+    })
+    .ElseIf("x > 5", () => new[]
+    {
+        "break;"
+    });
+
+    //logic
+});
+```
+
+## Miscellaneous
+
+### Define scopes as string
+
+trycatchfinally
```

Comment:
needs some description

Reply:
yeah this is where the `Func<IEnumerable<string>>` explanation will go